src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java

```diff
@@ -170,29 +197,68 @@
 	 */
 
 	private void pickBestCandidates(
-			HashMap<String, ArrayList<String>> resolvedEntities,
-			HashMap<String, ArrayList<ArrayList<String>>> allCandidates) {
+			HashMap<String, List<String>> resolvedEntities,
+			HashMap<String, List<List<String>>> allCandidates) {
 
 		for (String extractedName : allCandidates.keySet()) {
-			ArrayList<ArrayList<String>> cur = allCandidates.get(extractedName);
-			int minDistance = Integer.MAX_VALUE, minIndex = -1;
+			
+			List<List<String>> cur = allCandidates.get(extractedName);
+			if(cur.isEmpty())
+				continue;//continue if no results found
+			
+			int maxWeight = Integer.MIN_VALUE ;
+			//In case weight is equal for all return top element
+			int bestIndex = 0;
 			for (int i = 0; i < cur.size(); ++i) {
-				String resolvedName = cur.get(i).get(0);// get cur's ith
-														// resolved entry's name
-				int distance = StringUtils.getLevenshteinDistance(
-						extractedName, resolvedName);
-				if (distance < minDistance) {
-					minDistance = distance;
-					minIndex = i;
+				int weight;
+				// get cur's ith resolved entry's name
+				String resolvedName = cur.get(i).get(0);
+				//Assign a weight of 10 if extracted name is found in name
+				weight = resolvedName.contains(extractedName) ? 500 : 0;  
+				
+				// get all alternate names of cur's ith resolved entry's 
+				String[] altNames = cur.get(i).get(3).split(",");
+				float altEditDist = 0;
+				for(String altName : altNames){
+					if(altName.contains(extractedName)){
+						altEditDist+=StringUtils.getLevenshteinDistance(extractedName, altName);
+					}
+				}
+				//lesser the edit distance more should be the weight
+				weight += getCalibratedWeight(altNames.length, altEditDist);
+				
+				//Give preference to sorted results. 0th result should have more priority
+				weight += (cur.size()-i) * 2;
+						
+				if (weight > maxWeight) {
+					maxWeight = weight;
+					bestIndex = i;
 				}
 			}
-			if (minIndex == -1)
+			if (bestIndex == -1)
 				continue;
-			resolvedEntities.put(extractedName, cur.get(minIndex));
+			
+			//remove alternate name from allCandidates element before adding
+			cur.get(bestIndex).remove(3);
+			resolvedEntities.put(extractedName, cur.get(bestIndex));
 		}
 	}
 
 	/**
+	 * Returns a weight for average edit distance for set of alternate name<br/><br/>
+	 * altNamesSize * 50 - (altEditDist/altNamesSize) ;<br/><br/>
+	 * altNamesSize * 50 ensure more priority for results with more alternate names.<br/> 
+	 * altEditDist/altNamesSize is average edit distance. <br/>
+	 * Lesser the average, higher the over all expression
+	 * @param altNamesSize - Count of altNames
+	 * @param altEditDist - sum of individual edit distances
+	 * @return
+	 */
+	public float getCalibratedWeight(int altNamesSize, float altEditDist) { 
+		return altNamesSize * 50 - (altEditDist/altNamesSize) ;
```

Comment:
make 50 configurable and not a magic string - give it a name and make it a settable value for the class.